### PR TITLE
Support single stage rotated detector in MMRotate

### PR DESCRIPTION
**Pull request by zytx121**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily receiving feedbacks. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Support single stage rotated detector in MMRotate. (#312 is included.)

- [x] add mmrotate.deploy
- [x] add rewritings for mmrotate models
- [x] add mmrotate deploy configs
- [x] add unittests
- [x] add docs & update docs
- [x] align with pytorch

<div style="margin-left: 25px;">
<table class="docutils">
<thead>
  <tr>
    <th align="center" colspan="4">MMRotate</th>
    <th align="center">Pytorch</th>
    <th align="center">ONNXRuntime</th>
    <th align="center" colspan="2">TensorRT</th>
    <th align="center">PPLNN</th>
    <th align="center">OpenVINO</th>
    <th align="left">Model Config</th>
  </tr>
</thead>
<tbody>
  <tr>
    <td align="center">Model</td>
    <td align="center">Task</td>
    <td align="center">Dataset</td>
    <td align="center">Metrics</td>
    <td align="center">fp32</td>
    <td align="center">fp32</td>
    <td align="center">fp32</td>
    <td align="center">fp16</td>
    <td align="center">fp16</td>
    <td align="center">fp32</td>
    <td>model config file</td>
  </tr>
  <tr>
    <td align="center" rowspan="2">RotatedRetinaNet</td>
    <td align="center" rowspan="2">Rotated Detection</td>
    <td align="center" rowspan="2">DOTA-v1.0</td>
    <td align="center">mAP</td>
    <td align="center">0.698</td>
    <td align="center">0.698</td>
    <td align="center">-</td>
    <td align="center">-</td>
    <td align="center">-</td>
    <td align="center">-</td>
    <td rowspan="2">$MMROTATE_DIR/configs/rotated_retinanet/rotated_retinanet_obb_r50_fpn_1x_dota_le135.py</td>
  </tr>
</tbody>
</table>
</div>

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit tests to ensure the correctness.
3. If the modification has a dependency on downstream projects of a newer version, this PR should be tested with all supported versions of downstream projects.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: mmdeploy/codebase/mmrotate/deploy/rotated_detection.py [hunk 1]
```diff
@@ -0,0 +1,318 @@
...
+            model_checkpoint (str): The checkpoint file of torch model,
+                defaults to `None`.
+            cfg_options (dict): Optional config key-pair parameters.
+
+        Returns:
+            nn.Module: An initialized torch model generated by OpenMMLab
+                codebases.
+        """
+        from mmrotate.apis import init_detector
+        model = init_detector(self.model_cfg, model_checkpoint, self.device,
```
Comment (grimoire) [inline mmdeploy/codebase/mmrotate/deploy/rotated_detection.py:98]: There is no `init_detector` in `mmrotate.apis`.

File: mmdeploy/codebase/mmrotate/deploy/rotated_detection.py [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+                              cfg_options)
+
+        return model.eval()
+
+    def create_input(self,
+                     imgs: Union[str, np.ndarray],
+                     input_shape: Sequence[int] = None) \
+            -> Tuple[Dict, torch.Tensor]:
+        """Create input for segmentor.
```
Comment (grimoire) [inline mmdeploy/codebase/mmrotate/deploy/rotated_detection.py:107]: not segmentor

File: mmdeploy/codebase/mmrotate/deploy/rotated_detection.py [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+            if not isinstance(imgs[0], (np.ndarray, str)):
+                raise AssertionError('imgs must be strings or numpy arrays')
+
+        elif isinstance(imgs, (np.ndarray, str)):
+            imgs = [imgs]
+        else:
+            raise AssertionError('imgs must be strings or numpy arrays')
+        cfg = process_model_config(self.model_cfg, imgs, input_shape)
+        from mmdet.datasets.pipelines import Compose
+        from mmrotate.datasets import build_dataset  # noqa: F401
```
Comment (grimoire) [inline mmdeploy/codebase/mmrotate/deploy/rotated_detection.py:128]: Remove this if it has not been used.

File: mmdeploy/codebase/mmrotate/deploy/rotated_detection.py (hunk 1 above)
Comment (zytx121) [reply mmdeploy/codebase/mmrotate/deploy/rotated_detection.py:98]: Yes, maybe we should add init_detector in `mmrotate.apis`. I'll try if there's another way.

File: mmdeploy/mmcv/ops/nms_rotated.py [hunk 4]
```diff
@@ -13,22 +13,36 @@ def forward(ctx, boxes: Tensor, scores: Tensor,
...
-                [N, num_boxes, num_classes].
+                [N, num_classes, num_boxes].
             iou_threshold (float): IOU threshold of nms.
 
         Returns:
             Tensor: Selected indices of boxes.
         """
         from mmcv.utils import ext_loader
         ext_module = ext_loader.load_ext('_ext', ['nms_rotated'])
+        batch_size, num_class, _ = scores.shape
```
Comment (grimoire) [inline mmdeploy/mmcv/ops/nms_rotated.py:26]: deploy crashed here with error:
```
ValueError: not enough values to unpack (expected 3, got 1)
```

**Comment by hust-lidelong** [comment]
When to support two-stage rotated detector in MMRotate? Thanks

File: mmdeploy/codebase/mmrotate/core/post_processing/bbox_nms.py [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+    if pre_top_k > 0:
+        max_scores, _ = scores.max(-1)
+        _, topk_inds = max_scores.topk(pre_top_k)
+        batch_inds = torch.arange(batch_size).view(
+            -1, 1).expand_as(topk_inds).long()
+        boxes = boxes[batch_inds, topk_inds, :]
+        scores = scores[batch_inds, topk_inds, :]
+
+    scores = scores.permute(0, 2, 1)
+    selected_indices = ONNXNMSRotatedOp.apply(boxes, scores, iou_threshold,
```
Comment (grimoire) [inline mmdeploy/codebase/mmrotate/core/post_processing/bbox_nms.py:111]: `ONNXNMSRotatedOp` accept 3 parameters (`boxes: Tensor, scores: Tensor, iou_threshold: float`) but 4 inputs has been feeded here.

**Comment by zytx121** [comment]
> When to support two-stage rotated detector in MMRotate? Thanks

We will support two-stage rotated detector in MMRotate after this PR

File: mmdeploy/codebase/mmrotate/deploy/rotated_detection_model.py [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+        deploy_cfg: Union[str, mmcv.Config] = None,
+        model_cfg: Union[str, mmcv.Config] = None,
+    ):
+        super(End2EndModel, self).__init__(deploy_cfg=deploy_cfg)
+        model_cfg, deploy_cfg = load_config(model_cfg, deploy_cfg)
+        self.CLASSES = class_names
+        self.deploy_cfg = deploy_cfg
+        self.device = device
+        self.show_score = False
+        self.bbox_head = build_head(model_cfg.model.bbox_head)
```
Comment (grimoire) [inline mmdeploy/codebase/mmrotate/deploy/rotated_detection_model.py:57]: If this `self.bbox_head` has not been used, you can just remove it.

**Comment by zytx121** [comment]
It looks like boxes that score < 0.63 were filtered.

**Comment by zytx121** [comment]
> The visualize result is a little bit different from the PyTorch model. Some horizon boxes are missing.
> 

It is caused by the `pre_top_k` being too small.

**Review by grimoire — DISMISSED** [review]
LGTM

**Review by lzhangzz — DISMISSED** [review]

**Comment by zytx121** [comment]
Open new PR #427

**Pull request closed without merging**